### ISSUE TITLE
Update Hood to use generics

### ISSUE THREAD

**Issue by andygill** [ku-fpg/blackboard#7]
We may be able to replace Observable with Generic, in Hood, making automatic derivations much easier.

**Comment by RyanGlScott** [comment]
You mean [this](https://github.com/ku-fpg/hood/pull/1)? Or are you envisioning something more?

**Comment by andygill** [comment]
Yes! Thanks. I forgot that there was this pull request. There are some design choices to be considered here as well; before pushing to hackage, though. I'm sure @MaartenFaddegon will want to be part of any discussion. I was impressed with what he showed me at PLDI last summer, but do not remember any details.

**Comment by RyanGlScott** [comment]
> There are some design choices to be considered here as well; before pushing to hackage

Oops... I jumped the gun, then, since that's [already on Hackage](http://hackage.haskell.org/package/hood-0.3/changelog).

**Comment by andygill** [comment]
:-) No worries. I missed that the Generics went out with 0.3.

We should revisit the overall design, though, at some point.
- How can be make instance generation for regular data-structures trivial.
- How can be factor out the pretty printer, and make them plug-and-play, and perhaps support a web-based API.

**Comment by RyanGlScott** [comment]
I can at least answer the first question:

> How can be make instance generation for regular data-structures trivial.

This comes for extremely cheap, thanks to the power of `DefaultSignatures`. All you'd have to do is:

``` haskell
{-# LANGUAGE DeriveGeneric #-}
import Debug.Hood.Observe
import GHC.Generics

data Wibble a = Wibble Int Double [Wibble a] deriving Generic
instance Observable a => Observable (Wibble a)
```

Or, if using GHC 7.10 or later:

``` haskell
{-# LANGUAGE DeriveAnyClass, DeriveGeneric #-}
import Debug.Hood.Observe
import GHC.Generics

data Wibble a = Wibble Int Double [Wibble a] deriving (Generic, Observable)
```

**Comment by MaartenFaddegon** [comment]
The combinators from FPretty's API are shared between several pretty printing libraries (see e.g. wl-pprint). Hence swapping libraries can be done by just changing the import.

For the web interface, what do you have in mind? Maybe you want something a bit like the "Observe" view from Hoed?

![Screenshot of Hoed](https://github.com/MaartenFaddegon/Hoed/blob/master/screenshots/Observe.png)
